### PR TITLE
[fmiweather] HIRLAM → HARMONIE weather model

### PR DESCRIPTION
**Pull request by nils**
(Partial) bugfix attempt for the deprecation of the HIRLAM model provided my FMI.

https://community.openhab.org/t/fmi-weather-binding-hirlam-end-of-use/140642 and #13645

### CODE REVIEW

**Review by lolodomo — DISMISSED** [review]
Code LGTM

**Comment by lolodomo** [comment]
Maybe @ssalonen can confirm the bug fix and look at the change.

**Comment by lolodomo** [comment]
I certainly approved too fast as the PR is still work in progress.

**Comment by ssalonen** [comment]
Here is the news directly from source https://en.ilmatieteenlaitos.fi/news/yz6UzehhYMMzfCsMBTk3t

> HIRLAM weather forecast data will be removed from the Finnish Meteorological Institute's open data service on 1 November 2022. 

> In Northern Europe, the HARMONIE (MEPS) weather forecast data can be used as a replacement weather forecast model. The model predicts weather in the Finnish, Baltic and Scandinavian regions with a horizontal resolution of 2.5 kilometres 1-2 days away.


1) Readme still refers to HIRLAM and coverage of all Europe. 

With HARMONIE, we cover smaller area. (Map in https://en.ilmatieteenlaitos.fi/numerical-weather-prediction ?)

2) test resources refer to hirlam (see xml comment). Would be good to change that as well. I guess the output is the same format though (positive testing results in community forum)

**Comment by nils** [comment]
@ssalonen @lolodomo Please wait with a review until I move the PR out of draft. I appreciate a good review, but reviewing a draft PR is a waste of time for all of us. (Just as a fact, I'm not mad at anyone :slightly_smiling_face:)

**Comment by ssalonen** [comment]
thanks @nils  no offence taken :) 

This is how I analyzed the forecast horizon:
https://github.com/openhab/openhab-addons/pull/6329#issuecomment-618483730 (unfortunately do not have the analysis script anymore, it was piece of python/bash running in cron)

Apparently I used that to set some of the far future values under "advanced" flag

**Comment by ssalonen** [comment]
@nils how is it going with this one? I think new openHAB comes around year end so this would be excellent time to get this in still!

Really anything is better than currently broken version... :)

**Comment by nils** [comment]
Hi @ssalonen,

thanks for the reminder! I'm pretty much booked at the moment with other stuff, but I think this is good enough for the moment. The only thing missing that I'm aware of is the analysis of the time horizon:

> Btw, this is how I analyzed the forecast horizon:
> https://github.com/openhab/openhab-addons/pull/6329#issuecomment-618483730 (unfortunately do not have the analysis script anymore, it was piece of python/bash running in cron)

However I think that can be done separately, at least it looks like it is not very different from the previous model.

Would you please give the PR a review?

File: bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/CitiesOfFinland.java [hunk 1]
```diff
@@ -22,7 +22,7 @@
  * Cities of Finland
  *
  * Parsed from
- * https://opendata.fmi.fi/wfs?service=WFS&version=2.0.0&request=getFeature&storedquery_id=fmi::forecast::hirlam::surface::cities::multipointcoverage
+ * https://opendata.fmi.fi/wfs?service=WFS&version=2.0.0&request=getFeature&storedquery_id=fmi::forecast::harmonie::surface::cities::multipointcoverage
```
Comment (ssalonen) [inline bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/CitiesOfFinland.java:25]: Both URLs are not wrong. There seems to be no similar "stored query" cities.

Checking https://en.ilmatieteenlaitos.fi/open-data-manual-fmi-wfs-services we can find forecast for Finnish cities 

https://opendata.fmi.fi/wfs?service=WFS&version=2.0.0&request=getFeature&storedquery_id=ecmwf::forecast::surface::cities::multipointcoverage

But the list is different compared to currently hard-coded list.

Perhaps we just refer to old hirlam URL and mention that the resource is no longer working?

**Review by ssalonen — CHANGES_REQUESTED** [review]

**Comment by ssalonen** [comment]
> However I think that can be done separately, at least it looks like it is not very different from the previous model.

Yeah that can be done in separate PR.

I made small script to check this new behaviour

```bash
#!/bin/bash

NOW=$(date -u +"%Y-%m-%dT%H:%I:%MZ")
# current time floored down to even hour
START=$(date -u +"%Y-%m-%dT%H:00:00Z")
# current time + 80 hours, rounded to even hour
END=$(date --date "80 hours" -u +"%Y-%m-%dT%H:00:00Z")

# Calculate number of Not-a-Number values
NANS=$(curl --silent "https://opendata.fmi.fi/wfs?service=WFS&version=2.0.0&request=getFeature&storedquery_id=fmi::forecast::harmonie::surface::point::multipointcoverage&starttime=$START&endtime=$END&timestep=10&latlon=60.192059,24.945831&parameters=Temperature" | grep NaN | wc -l)
# Use the Nan-Count to determine number of valid values
NONNANS=$(( (80 * 6 - $NANS) / 6 ))

# Log to file
echo "$NOW;$NONNANS">>/home/pi/horizons.txt
```

**Comment by ssalonen** [comment]
LGTM

**Review by ssalonen — APPROVED** [review]

**Comment by lolodomo** [comment]
Please note there is a conflict to solve with README.

**Comment by lolodomo** [comment]
Does this PR still require additional testing or not ?

**Comment by ssalonen** [comment]
@nils please see above the forecast horizon for HARMONIE https://github.com/openhab/openhab-addons/pull/13693#issuecomment-1336400256

I think we should add additional channels [here](https://github.com/openhab/openhab-addons/blob/db86d291dab319bfd7449cf2b823690d38d56a74/bundles/org.openhab.binding.fmiweather/src/main/resources/OH-INF/thing/thing-types.xml#L31-L234)

- forecastNow, forecastHours01, forecastHours02, ..., forecastHours59 as normal channels, 
- forecastHours59, forecastHours60, ..., forecastHours63 as _advanced_ channels

In addition, README should be slightly updates (reference to 44 h [here](https://github.com/openhab/openhab-addons/blame/main/bundles/org.openhab.binding.fmiweather/README.md#L82)), and the examples as well

This can be done in a separate PR though. It looks like the advanced status of group have been removed by @J-N-K in [this commit](https://github.com/openhab/openhab-addons/commit/6616353baaf9aefa1514c18cc5a86a6b4be3bad6 ) (as a result of https://github.com/openhab/openhab-core/pull/1410), but without introducing the `advanced` flag to channels . 

All channels in group `group-forecast-advanced` should probably  have the advanced flag enabled, as commented by @ghys in https://github.com/openhab/openhab-core/issues/1402#issuecomment-609074097 

> ...you can easily make an advanced group by flagging all of its channels as advanced...

**Comment by ssalonen** [comment]
@nils ?

**Comment by nils** [comment]
@lolodomo  @ssalonen Please excuse the delay. I have fixed the conflict, apart from that nothing has been changed.

From my perspective it is ready to merge. As my availablities won't improve in the next weeks, I'd like to ask @ssalonen to do the follow-up changes required for the horizon adjustment.

**Comment by ssalonen** [comment]
@lolodomo can this still make the release? Current version of fmi weather is completely broken so it would be great to get this

**Comment by ssalonen** [comment]
Would this fit the 3.4.x maintenance branch @lolodomo @nils ?

**Review by lolodomo — APPROVED** [review]
LGTM, I have not checked in details all the changes in the documentation example (but it seems to be a simple renaming of Paris into Helsinki).

**Comment by lolodomo** [comment]
> Would this fit the 3.4.x maintenance branch @lolodomo @nils ?

Sorry for the delay, the PR looks very big but it was in fact mainly items renaming in the documentation.

As the binding is fully broken in 3.4, yes I would suggest to make this PR also in branch 3.4.x. @jlaur: can you do it please ? ;)

**Pull request merged**